### PR TITLE
New user group membership management

### PR DESCRIPTION
**Pull request by ksuess**
Merged and with **plone.restapi 8.24.0** released changes:
- Merged: https://github.com/plone/plone.restapi/pull/1325 
- Merged: https://github.com/plone/plone.restapi/pull/1439 
- Merged: For testing the many_users/many_groups case https://github.com/plone/plone.restapi/pull/1435 is needed.
- Merged: Search for fullname: https://github.com/plone/plone.restapi/pull/1444
   
----------------------------

grafik

  
Fix https://github.com/plone/volto/issues/3018

### CODE REVIEW

**Comment by ksuess** [comment]
The draft is ready for reviewing. Before I switch to 'ready': 
These PRs need review:
@tisto @sneridagh please review https://github.com/plone/plone.restapi/pull/1325 and https://github.com/plone/plone.restapi/pull/1435, thanks!

File: src/actions/users/users.js [hunk 1]
```diff
@@ -69,12 +69,20 @@ export function getUser(id) {
...
  * @returns {Object} List users action
  */
-export function listUsers(query) {
+export function listUsers(query, groups_filter) {
+  groups_filter = groups_filter || [];
+  let path = '/@users';
+  if (query) {
+    path += `?query=${query}`;
+    if (groups_filter.length > 0) {
+      path += `&groups_filter=${groups_filter.join(',')}`;
```
Comment (nileshgulia1) [inline src/actions/users/users.js:78]: @ksuess I think we should also have to change this `groups_filter` as per https://github.com/plone/plone.restapi/commit/bc42a349bb7740d7a61155d4f0c50c92a66c5c9f 🤔

Comment (ksuess) [reply src/actions/users/users.js:78]: Yes of course, thank you @nileshgulia1

**Comment by erral** [comment]
I reviewed this and it's really an impressive work, congrats!

I suggest 2 small improvements, feel free to comment on them:

- Somehow obey the "Many users" setting of Plone. This way all users will not be shown in the panel if the setting is enabled
- Allow searching also in the fullname, right now only username based searches are allowed

**Comment by ksuess** [comment]
> I reviewed this and it's really an impressive work, congrats!
> 
> I suggest 2 small improvements, feel free to comment on them (or just discard them 👅 ):
> 
>     * Somehow obey the "Many users" setting of Plone. This way all users will not be shown in the panel if the setting is enabled
> 
>     * Allow searching also in the fullname, right now only username based searches are allowed

Thanks for your review and feedback @erral!

The options "many_user" and "many_groups" are obeyed. The requested changes for this PR to plone.restapi were split in three PRs. Two are merged but not released.
There is one left for @plone/volto-team :  https://github.com/plone/plone.restapi/pull/1325
Check out https://github.com/plone/plone.restapi/pull/1325 for switching options "many_user" and "many_groups".

<  searching for fullname
hm, i do want this. didn't i implemented it?

**Comment by erral** [comment]
ah OK i tested just with one branch of plone.restapi, let me check it.

Regarding the fullname, i tested and I only could search for username, strange... The search is done in the front or backend?

**Comment by erral** [comment]
I was testing with that exact branch, and I don't see any difference enabling or disabling the `many_users` option, am I missing something?

**Comment by erral** [comment]
Ah... I see, the PR is not rebased, so I wasn't testing it with the latest PRs merged

**Comment by erral** [comment]
It looks like the controlpanel serializer fails in my Plone site when calling the `@usergroup` endpoint and that's the reason why I wasn't seeing the many_users and many_groups stuff.

**Comment by nileshgulia1** [comment]
>It looks like the controlpanel serializer fails in my Plone site when calling the @controlpanels/usergroup endpoint and that's the reason why I wasn't seeing the many_users and many_groups stuff.

We do have a eionet LDAP at EEA where we can test many_users flag. I will try to test this branch today.

**Review by nileshgulia1 — COMMENTED** [review]
Screenshot 2022-06-08 at 4 52 16 PM
@ksuess I just tested it a bit with our LDAP account and it seems that the request to `@groups` and `@users`, when enabled those flags even through the plone-classic, doesn't have an effect. This also happens in the normal users listing in usercontrolpanel ("/controlpanel/users").

Perhaps I could propose that we can protect those endpoints when `many_users/many_groups` turned on to return no data until a user performs a search.

**Comment by erral** [comment]
FWIW I was testing with the buildout provided by plone.restapi itself which is based on Plone 5.2.x

**Comment by ksuess** [comment]
@erral 
> It looks like the controlpanel serializer fails in my Plone site when calling the `@controlpanels/usergroup` endpoint and that's the reason why I wasn't seeing the many_users and many_groups stuff.
> 
> Specifically the usergroup controlpanel does not have a title and group attributes and it fails when rendering it
> 

Sorry, I can't reproduce. In the meantime further fixes were done, so maybe this is already solved. 
The settings "many users" and "many groups" are taken from `@controlpanels/usergroup`. And cause its schema `IUserGroupsSettingsSchema` is already in Plone 5, this is OK also for Plone 5, as long as you checkout for now plone.restapi master until released. 

BTW I take care to update the description of this PR with every step. Multiple PRs on plone.restapi are involved.

> Regarding the user search, the `@users` endpoint just searches based on `id`:
> 
> https://github.com/plone/plone.restapi/blob/bc42a349bb7740d7a61155d4f0c50c92a66c5c9f/src/plone/restapi/services/users/get.py#L55

This is pipelined. Please review: https://github.com/plone/plone.restapi/pull/1444

> FWIW I was testing with the buildout provided by plone.restapi itself which is based on Plone 5.2.x

Hm, I checked on both Plone 6 and Plone 5 just now again. Please comment again, if you still have problems.

Thank you for the feedback!

**Comment by ksuess** [comment]
I reworked the panel for many users and many groups
- No requests to the backend, until you search or filter.
- page size is the default of the backend: 25. To see more, a button at the end of the list requests one more batch per click. No old style pagination.

@nileshgulia1, @avoinea, please throw an eye on this.

Please review also the last of four dependencies in plone.restapi: https://github.com/plone/plone.restapi/pull/1444 to finish the _user group membership control panel_ . This would also help the CI here.

image

**Review by nileshgulia1 — COMMENTED** [review]
@ksuess I reviewed the PR with both LDAP and plone-buildout and it all looks great. 
One small thing, Is it possible to dedupe the `@groups` requests when the `many_groups`flag is off?
Screenshot 2022-06-27 at 4 14 40 PM

**Review by nileshgulia1 — APPROVED** [review]
LGTM 🎉

**Comment by nileshgulia1** [comment]
@ksuess Also Let's wait for @sneridagh  and we can merge after cypress is green!

**Comment by ksuess** [comment]
Please release plone.restapi. Tests are prepared. I would love to see them green!

**Comment by ksuess** [comment]
Group membership control panel is a need in a project. @tisto, please release plone.restapi to get tests green here, thanks.

**Comment by ksuess** [comment]
Wouldn't  it be cool to have a panel that supports adding users to groups? Here it is!
You're just a review step away! Tests are greener than green, cucumber green!

File: src/components/manage/Controlpanels/Users/UserGroupMembershipControlPanel.jsx [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    (state) => state.controlpanels.controlpanel?.data?.many_users,
+  );
+  const many_groups = useSelector(
+    (state) => state.controlpanels.controlpanel?.data?.many_groups,
+  );
+  const systeminformation = useSelector(
+    (state) => state.controlpanels.systeminformation,
+  );
+  const i_can_use_group_membership_panel = systeminformation
+    ? systeminformation.plone_version[0] >= 6 &&
```
Comment (erral) [inline src/components/manage/Controlpanels/Users/UserGroupMembershipControlPanel.jsx:41]: Why are we requiring Plone 6 here? I have removed this check and the control panel works as expected in Plone 5.2

**Review by erral — APPROVED** [review]
Apart from the Plone 5.2 vs. Plone 6 check this works as expected.

File: src/components/manage/Controlpanels/Users/UserGroupMembershipControlPanel.jsx (hunk 2 above)
Comment (ksuess) [reply src/components/manage/Controlpanels/Users/UserGroupMembershipControlPanel.jsx]: Can't reproduce anymore. I thought it was a problem with acl_users.searchUsers(id=query) with query == None. But I stepped back to Plone 5.2.7 and the problem is not reproducable.
So restriction can be removed.

File: src/components/manage/Controlpanels/Controlpanels.jsx [hunk 3]
```diff
@@ -167,6 +167,11 @@ class Controlpanels extends Component {
           group: this.props.intl.formatMessage(messages.usersandgroups),
           title: this.props.intl.formatMessage(messages.users),
         },
+        {
+          '@id': '/usergroupmembership',
+          group: 'Users and Groups',
+          title: 'User Group Membership', // this.props.intl.formatMessage(messages.users),
```
Comment (sneridagh) [inline src/components/manage/Controlpanels/Controlpanels.jsx:173]: Can we re-add the i18n config? We have an undergoing PR trying to fix some of the other CP translations already: https://github.com/plone/volto/pull/3486/files#diff-9d05cfbd5917826f55a5865e35ea15138f8054bf5d66f95192151af7586cdd1d

**Review by sneridagh — APPROVED** [review]
LGTM, take a look at my comment, or address it in another PR.

Thanks! amazing work!

File: src/components/manage/Controlpanels/Controlpanels.jsx (hunk 3 above)
Comment (ksuess) [reply src/components/manage/Controlpanels/Controlpanels.jsx:173]: oops yes, of course. Will do today.
Thanks for taking time and reviewing!

Comment (ksuess) [reply src/components/manage/Controlpanels/Controlpanels.jsx]: Thanks for hte hint! Done in https://github.com/plone/volto/pull/3349/commits/70ce484e65deadc0865357bee8de0caaacdadfe5

**Pull request merged**